### PR TITLE
[DAG] SimplifyMultipleUseDemandedBits - ignore SRL node if we're just demanding known sign bits

### PR DESCRIPTION
**Pull request by RKSimon**
Check to see if we are only demanding (shifted) signbits from a SRL node that are also signbits in the source node.

We can't demand any upper zero bits that the SRL will shift in (up to max shift amount), and the lower demanded bits bound must already be all signbits.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 1]
```diff
@@ -808,6 +808,24 @@ SDValue TargetLowering::SimplifyMultipleUseDemandedBits(
...
+    // If we are only demanding sign bits then we can use the shift source
+    // directly.
+    if (std::optional<uint64_t> MaxSA =
+            DAG.getValidMaximumShiftAmount(Op, DemandedElts, Depth + 1)) {
+      SDValue Op0 = Op.getOperand(0);
+      unsigned ShAmt = *MaxSA;
+      unsigned NumSignBits =
+          DAG.ComputeNumSignBits(Op0, DemandedElts, Depth + 1);
+      unsigned LoDemandedBits = DemandedBits.countr_zero();
+      unsigned HiDemandedBits = DemandedBits.countl_zero();
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:821]: Nit: aren't these more like Lo**Un**demandedBits and Hi**Un**demandedBits?

**Review by jayfoad — APPROVED** [review]
Makes sense to me. One naming nit inline.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 1 above)
Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:821]: Sure - they were original LoBoundDemandedBits but that caused clang-format wrapping :)

**Review by jayfoad — APPROVED** [review]

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 2]
```diff
@@ -808,6 +808,23 @@ SDValue TargetLowering::SimplifyMultipleUseDemandedBits(
...
+    // directly.
+    if (std::optional<uint64_t> MaxSA =
+            DAG.getValidMaximumShiftAmount(Op, DemandedElts, Depth + 1)) {
+      SDValue Op0 = Op.getOperand(0);
+      unsigned ShAmt = *MaxSA;
+      unsigned NumSignBits =
+          DAG.ComputeNumSignBits(Op0, DemandedElts, Depth + 1);
+      // Must already be signbits in DemandedBits bounds, and can't demand any
+      // shifted in zeroes.
+      if (DemandedBits.countl_zero() >= ShAmt &&
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:822]: nit: guard the `ComputeNumSignBits` behind this check.

**Pull request merged**